pkg/k8sops/generators.go

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package k8sops
+
+import (
+	"errors"
+	"fmt"
+
+	m3dboperator "github.com/m3db/m3db-operator/pkg/apis/m3dboperator"
+	myspec "github.com/m3db/m3db-operator/pkg/apis/m3dboperator/v1"
+
+	appsv1 "k8s.io/api/apps/v1"
+	"k8s.io/api/core/v1"
+	apiextensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/intstr"
+)
+
+// GenerateCRD generates the crd object needed for the M3DBCluster
+func (k *K8sops) GenerateCRD() *apiextensionsv1beta1.CustomResourceDefinition {
+	return &apiextensionsv1beta1.CustomResourceDefinition{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: m3dboperator.Name,
+		},
+		Spec: apiextensionsv1beta1.CustomResourceDefinitionSpec{
+			Group:   m3dboperator.GroupName,
+			Version: m3dboperator.Version,
+			Scope:   apiextensionsv1beta1.NamespaceScoped,
+			Names: apiextensionsv1beta1.CustomResourceDefinitionNames{
+				Plural: m3dboperator.ResourcePlural,
+				Kind:   m3dboperator.ResourceKind,
+			},
+		},
+	}
+}
+
+// GenerateService generates the Service object for m3coordinator
+func (k *K8sops) GenerateService(name string) (*v1.Service, error) {
+
+	switch name {
+	case M3dbNodeSvcName:
+		ports, err := k.GenerateServicePorts(name)
+		if err != nil {
+			return nil, err
+		}
+		return &v1.Service{
```

Comment:
not familiar at all with the k8 apis, but this whole file feels like it's trying to write some config yaml in go - any way of simplifying the code?

Reply:
So I was able to convert some of the file into dynamic conversion. There a plenty with this though, if you take a look at `manifests/configmap.yaml` which contains `configmap`s of the configuration which need to be mounted at file locations in which Kubernetes is strict about unique mount points e.g. can not have the same directory mapped to a configmap which adds overhead to the operator's mounted points. Furthermore if any of these configurations contain nested configurations such as `ContainerPort` and `ServicePort` in which we would need to reconcile those two objects together. If you look at `LoadService` will show that issue since we expect configmap objects to be wholesale and not templated. IHMO it will be over optimizing the initial delivery of the operator. I agree these parameters shouldn't require a redeploy of the operator however we can make that a v2 feature.

Reply:
I was able to accomplish this via the use of configmaps and marshaling into the actual structs themselves. However that approach is over complicated. After talking with @schallert , a better approach would be to have a select amount of configuration parameters exposed the spec of the M3DBCluster object and use those values if present otherwise use defaults. This would apply to service names, service ports, and container ports.

Reply:
Nice, I like it. At some point I can see us generating these configs to configure the ports or w/e, will be much easier in the structure you have now.